### PR TITLE
Support Endpoints for Instance Datums

### PR DESCRIPTION
**Pull request by shubham1g5**
See https://github.com/dimagi/formplayer/pull/1431

### Special deploy instructions

- Corresponding FP PR will need to be reverted if we revert this. 

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations.

### Review

- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change.

### Duplicate PR
Automatically duplicate this PR as defined in [contributing.md](https://github.com/dimagi/commcare-core/blob/master/.github/contributing.md).

### CODE REVIEW

File: src/main/java/org/commcare/xml/EndpointParser.java [hunk 1]
```diff
@@ -31,16 +35,23 @@ public Endpoint parse() throws InvalidStructureException, IOException, XmlPullPa
...
                 }
-                arguments.add(argumentID);
+                String argInstanceId = parser.getAttributeValue(null, ATTR_ARGUMENT_INSTANCE_ID);
+                String argInstanceSrc = parser.getAttributeValue(null, ATTR_ARGUMENT_INSTANCE_SRC);
+
+                if (argInstanceId != null && argInstanceSrc == null) {
+                    throw new InvalidStructureException(
+                            "argument containing a non-null instance-id must define an instance-src", parser);
+                }
+                arguments.add(new EndpointArgument(argumentID, argInstanceId, argInstanceSrc));
```
Comment (Robert-Costello) [inline src/main/java/org/commcare/xml/EndpointParser.java:64]: Just to make sure I understand correctly, is the idea here that `EndpointArgument` also handles single case selection where `<argument id="case_id"/>`? So if there is no `argInstanceId ` and no `argInstanceSrc ` the functionality from this point is essentially the same as before.

Comment (shubham1g5) [reply src/main/java/org/commcare/xml/EndpointParser.java:64]: yes, spot on.

File: src/main/java/org/commcare/suite/model/Endpoint.java [hunk 2]
```diff
@@ -59,40 +60,60 @@ public Vector<StackOperation> getStackOperations() {
...
     // Utility Functions
     public static void populateEndpointArgumentsToEvaluationContext(Endpoint endpoint, ArrayList<String> args, EvaluationContext evaluationContext) {
-        Vector<String> endpointArguments = endpoint.getArguments();
+        Vector<EndpointArgument> endpointArguments = endpoint.getArguments();
 
         if (endpointArguments.size() > args.size()) {
-            Vector<String> missingArguments = new Vector<String>(endpointArguments.subList(args.size(), endpointArguments.size()));
+            Vector<String> missingArguments = new Vector<>();
+            for (int i = args.size(); i < endpointArguments.size(); i++) {
+                missingArguments.add(endpointArguments.get(i).getId());
```
[the remark this answers is not in the document]
Comment (shubham1g5) [reply src/main/java/org/commcare/suite/model/Endpoint.java:68]: Yeah we expect the arguments to be passed in order when it's passed as a list without using the argument keys (`arg1=val1&arg2=val2` vs `val1,val2`). Web Apps uses the fuirst format with names but Android can use both those formats. The ordering matters only in the later format.

**Review by Robert-Costello — APPROVED** [review]

File: src/cli/java/org/commcare/util/screen/MultiSelectEntityScreen.java [hunk 3]
```diff
@@ -134,9 +135,18 @@ private void prcessSelectionAsGuid(String guid) throws CommCareSessionException
                     "Could not make selection with reference id " + guid + " on this screen. " +
                             " If this error persists please report a bug to CommCareHQ.");
         }
+        validateEntitiesInInstance(cachedInstance);
         storageReferenceId = guid;
     }
 
+    private void validateEntitiesInInstance(ExternalDataInstance instance) throws CommCareSessionException {
```
Comment (shubham1g5) [inline src/cli/java/org/commcare/util/screen/MultiSelectEntityScreen.java:142]: There is currently[ an issue](https://dimagi-dev.atlassian.net/browse/QA-5457) with this validation that I am verifying with delivery

**Comment by shubham1g5** [comment]
duplicate this PR

**Pull request merged**